library/Icingadb/View/BaseHostAndServiceRenderer.php

```diff
@@ -0,0 +1,292 @@
+                        'div',
+                        $commentItem->getAttributes()->add('class', 'comment-popup'),
+                        $commentItem
+                    ),
+                    (new Icon('comments', ['class' => 'comment-icon']))
+                )
+            );
+        }
+
+        if ($item->state->is_flapping) {
+            $title = $isService
+                ? sprintf(
+                    $this->translate('Service "%s" on "%s" is in flapping state'),
+                    $item->display_name, $item->host->display_name
+                )
+                : sprintf(
+                    $this->translate('Host "%s" is in flapping state'),
+                    $item->display_name
+                );
+
+            $statusIcons->addHtml(new Icon('random', ['title' => $title]));
+        }
+
+        if (! $item->notifications_enabled) {
+            $statusIcons->addHtml(
+                new Icon('bell-slash', ['title' => $this->translate('Notifications disabled')])
+            );
+        }
+
+        if (! $item->active_checks_enabled) {
+            $statusIcons->addHtml(
+                new Icon('eye-slash', ['title' => $this->translate('Active checks disabled')])
+            );
+        }
+
+        $performanceData = new HtmlElement('div', Attributes::create(['class' => 'performance-data']));
+        if ($item->state->performance_data) {
+            $pieChartData = PerfDataSet::fromString($item->state->normalized_performance_data)->asArray();
+
+            $pies = [];
+            foreach ($pieChartData as $i => $perfdata) {
+                if ($perfdata->isVisualizable()) {
+                    $pies[] = $perfdata->asInlinePie()->render();
+                }
+
+                // Check if number of visualizable pie charts is larger than $PIE_CHART_LIMIT
+                if (count($pies) > $pieChartLimit) {
+                    break;
+                }
+            }
+
+            $maxVisiblePies = $pieChartLimit - 2;
+            $numOfPies = count($pies);
+            foreach ($pies as $i => $pie) {
+                if (
+                    // Show max. 5 elements: if there are more than 5, show 4 + `…`
+                    $i > $maxVisiblePies && $numOfPies > $pieChartLimit
+                ) {
+                    $performanceData->addHtml(new HtmlElement('span', null, Text::create('…')));
+                    break;
+                }
+
+                $performanceData->addHtml(HtmlString::create($pie));
+            }
+        }
+
+        if (! $statusIcons->isEmpty()) {
+            $footer->addHtml($statusIcons);
+        }
+
+        if (! $performanceData->isEmpty()) {
+            $footer->addHtml($performanceData);
+        }
+    }
+
+    public function assemble($item, string $name, HtmlDocument $element, string $layout): bool
+    {
+        if ($name === 'icon-image') {
+            $element->addHtml(new IconImage($item->icon_image->icon_image, $item->icon_image_alt));
```

Comment:
Not all items have an icon image and assembling the section does not guarantee it either. The previous implementation verified this.